File: src.old/classes/MathClasses/Quaternion.lua
---@class Quaternion 
---@field x number
---@field y number
---@field z number
---@field w number
---@field fromAdvancedLib fun(original:any):Quaternion
---@field multiply fun(self:self, vec:Vector3):Vector3
---@field fromAxisAngle fun(vector:Vector3, angle:number):Quaternion
---@field ConvertVectorToWorldSpace fun(self:self, vec:Vector3):Vector3
---@field ConvertVectorToLocalSpace fun(self:self, vec:Vector3):Vector3
---@field Inverse fun():Quaternion
---@field new fun(x:number, y:number, z:number, w:number):Quaternion

local std = require("classes.Standard");
---@type Vector3
local Vector3 = require("classes.MathClasses.Vector3");


local Quaternion = {
    _type = "Quaternion"
}
Quaternion.__index = Quaternion;

---@param quat Quaternion
---@param vec Vector3
Quaternion.__mul = function(quat, other)
    return quat:multiply(quat, other);
end

---@param vector Vector3
---@param angle number
function Quaternion.fromAxisAngle(vector, angle)
    local len = math.sqrt(vector.x^2 + vector.y^2 + vector.z^2)

    if len < 1e-8 then
        return Quaternion.New(0, 0, 0, 1)
    end

    local x = vector.x / len
    local y = vector.y / len
    local z = vector.z / len

    local half = angle * 0.5
    local s = math.sin(half)
    local c = math.cos(half)

    return Quaternion.New(
        x * s,
        y * s,
        z * s,
        c
    )
end

---@return Quaternion
function Quaternion:Inverse()
   local squaredNorm = self.x^2 + self.y^2 + self.z^2 + self.w^2;
   local inverted = Quaternion.New(-self.x/squaredNorm, -self.y/squaredNorm, -self.z/squaredNorm, self.w/squaredNorm);
   return inverted;
end

---@param v Vector3|Quaternion;
function Quaternion:multiply(this, v)
    -- Generated this function with the big GPT going to be so for real I have a love hate relationship with quaternions
    if std.typeof(v) == "Vector3" then
        local x, y, z = self.x, self.y, self.z;
        local w = self.w;

        -- t = 2 * cross(q.xyz, v)
        local tx = 2 * (y * v.z - z * v.y);
        local ty = 2 * (z * v.x - x * v.z);
        local tz = 2 * (x * v.y - y * v.x);

        -- v + w * t
        local vx = v.x + w * tx;
        local vy = v.y + w * ty;
        local vz = v.z + w * tz;

        -- + cross(q.xyz, t)
        vx = vx + (y * tz - z * ty);
        vy = vy + (z * tx - x * tz);
        vz = vz + (x * ty - y * tx);

        local endVector = Vector3.New(vx, vy, vz);
        return endVector;
    elseif std.typeof(v) == "Quaternion" then
        local x = self.w*v.x + self.x*v.w + self.y*v.z - self.z*v.y;
        local y = self.w*v.y - self.x*v.z + self.y*v.w + self.z*v.x;
        local z = self.w*v.z + self.x*v.y - self.y*v.x + self.z*v.w;
        local w = self.w*v.w - self.x*v.x - self.y*v.y - self.z*v.z;
        return Quaternion.New(x,y,z,w);
    end
    
end

---@param vec Vector3 
--- Converts a vector to world space, assuming that the rotation of the vector can be expressed by this quaternion
function Quaternion:ConvertVectorToWorldSpace(vec)
    local vecQuat = Quaternion.New(vec.x, vec.y, vec.z, 0);
    local intermediate = self * vecQuat;
    local finalQuat = intermediate * self:Inverse();
    local final = Vector3.New(finalQuat.x, finalQuat.y, finalQuat.z);
    return final;
end
---@param vec Vector3 
--- Converts a vector to local space
function Quaternion:ConvertVectorToLocalSpace(vec)
    local vecQuat = Quaternion.New(vec.x, vec.y, vec.z, 0);
    local intermediate = self:Inverse() * vecQuat;
    local finalQuat = intermediate * self;
    local final = Vector3.New(finalQuat.x, finalQuat.y, finalQuat.z);
    return final;
end
function Quaternion.New(x, y, z, w)
    local instance = {};
    instance.x = x;
    instance.y = y;
    instance.z = z;
    instance.w = w;
    setmetatable(instance, Quaternion);
    return instance;
end
function Quaternion.fromAdvancedLib(original)
    local instance = {};
    setmetatable(instance, Quaternion);
    instance.x = original.v.x;
    instance.y = original.v.y;
    instance.z = original.v.z;
    instance.w = original.a;
    return instance
end

return Quaternion;

File: src/classes/Math/Quaternion.ts
import { Vector3 } from "./Vector3";

export class Quaternion {
    x : number = 0;
    y : number = 0;
    z : number = 0;
    w : number = 0;

    constructor(wPassthrough : number, xPassthrough : number, yPassthrough : number, zPassthrough : number)
    {
        this.x = xPassthrough;
        this.y = yPassthrough;
        this.z = zPassthrough;
        this.w = wPassthrough;
    }

    static fromAxisAngles(axis : Vector3, angle : number): Quaternion
    {
        const mag : number = axis.Magnitude();

        if (mag < 1e-8) {
            let quat : Quaternion = new Quaternion(1,0,0,0);
            return quat;
        }

        const normalized : Vector3 = axis.Normalized();
        const halfAngle : number = angle * 0.5;

        const sinAngle : number = Math.sin(halfAngle);
        const cosAngle : number = Math.cos(halfAngle);

        let finalQuat : Quaternion = new Quaternion(cosAngle, normalized.x * sinAngle, normalized.y * cosAngle, normalized.z * cosAngle);
        return finalQuat;
    }

    Inverse(this : Quaternion): Quaternion
    {
       const squaredNorm : number = this.x ** 2 + this.y ** 2 + this.z ** 2 + this.w ** 2;
       const inverted : Quaternion = new Quaternion(-this.w/squaredNorm, -this.x/squaredNorm, -this.y/squaredNorm, -this.z/squaredNorm);
       return inverted;
    }

    MultiplyByVector(this:Quaternion, vector:Vector3) : Vector3
    {
        const tx : number = 2 * (this.y * vector.z - this.z * vector.y);
        const ty : number = 2 * (this.z * vector.x - this.x * vector.z);
        const tz : number = 2 * (this.x * vector.y - this.y * vector.x);

        const vx : number = vector.x + this.w * tx;
        const vy : number = vector.y + this.w * ty;
        const vz : number = vector.z + this.w * tz;
        /*
            vx = vx + (y * tz - z * ty);
            vy = vy + (z * tx - x * tz);
            vz = vz + (x * ty - y * tx);
        */
        const endVector : Vector3 = new Vector3(
            vx + (this.y * tz - this.z * ty),
            vy + (this.z * tx - this.x * tz),
            vz + (this.x * ty - this.y * tx)
        )
        return endVector;
    }

    MultiplyByQuaternion(this : Quaternion, other : Quaternion) : Quaternion
    {
        const x : number = this.w * other.x + this.x * other.w + this.y * other.z - this.z * other.y;
        const y : number = this.w * other.y - this.x * other.z + this.y * other.w + this.z * other.x;
        const z : number = this.w * other.z + this.x * other.y - this.y * other.x + this.z * other.w;
        const w : number = this.w * other.w - this.x * other.x - this.y * other.y - this.z * other.z;
        const endQuaternion : Quaternion = new Quaternion(w,x,y,z);
        return endQuaternion;
    }
   ConvertVectorToWorldSpace(this : Quaternion, vector : Vector3) : Vector3
   {
        const quatedVector : Quaternion = new Quaternion(0, vector.x, vector.y, vector.z);
        const intermediate : Quaternion = this.MultiplyByQuaternion(quatedVector);
        const finalQuat : Quaternion = intermediate.MultiplyByQuaternion(this.Inverse());
        const final : Vector3 = new Vector3(finalQuat.x, finalQuat.y, finalQuat.z);
        return final;
   }

   ConvertVectorToLocalSpace(this : Quaternion, vector : Vector3) : Vector3
   {
        const quatedVector : Quaternion = new Quaternion(0, vector.x, vector.y, vector.z);
        const intermediate : Quaternion = this.Inverse().MultiplyByQuaternion(quatedVector);
        const finalQuat : Quaternion = intermediate.MultiplyByQuaternion(this);
        const final : Vector3 = new Vector3(finalQuat.x, finalQuat.y, finalQuat.z);
        return final;
   }
}